### PR TITLE
Fix pre-1.0 patch versions

### PR DESCRIPTION
**Pull request by iskakaushik**
fix: https://github.com/flutter/flutter/issues/35411

### CODE REVIEW

File: lib/src/version_check_command.dart [hunk 1]
```diff
@@ -86,6 +86,24 @@ class VersionCheckCommand extends PluginCommand {
...
+            final String currentBuildNumber = masterVersion.build.first;
+            nextBuildNumber = int.parse(currentBuildNumber) + 1;
+          }
+          final Version preReleaseVersion = Version(
+            masterVersion.major,
+            masterVersion.minor,
+            masterVersion.patch,
+            build: nextBuildNumber.toString(),
+          );
+          allowedNextVersions[preReleaseVersion] = "PRE-1.0-PATCH";
```
Comment (amirh) [inline lib/src/version_check_command.dart]: This should say PATCH.
I believe the allowed versions message will be wrong here, e.g for `0.1.2+3` it should say:
Release: 1.0.0
Major: 0.2.0
Minor: 0.1.3
Patch: 0.1.2+4

Comment (iskakaushik) [reply lib/src/version_check_command.dart]: Done.

**Comment by iskakaushik** [comment]
@amirh, cool, will add some tests.

**Comment by iskakaushik** [comment]
Added comprehensive tests at https://github.com/flutter/plugin_tools/pull/38/files#diff-eeff3192fae4b2e81d5d275dc236d405

File: lib/src/version_check_command.dart [hunk 2]
```diff
@@ -41,6 +42,47 @@ class GitVersionFinder {
...
+enum NextVersionType {
+  BREAKING_MAJOR,
+  MINOR,
+  PATCH,
+  RELEASE,
+}
+
+@visibleForTesting
+Map<Version, NextVersionType> getAllowedNextVersions(
+    Version masterVersion, Version headVersion) {
```
Comment (amirh) [inline lib/src/version_check_command.dart]: Isn't the list of allowed versions just a function of the head version?

File: test/version_check_test.dart [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  } else {
+    expect(allowedVersions, isNot(contains(head)));
+  }
+}
+
+void main() {
+  group("Pre 1.0", () {
+    test("nextVersion allows patch version", () {
+      testAllowedVersion("0.12.0", "0.12.0+1",
```
Comment (amirh) [inline test/version_check_test.dart]: Landing straight at this test case it's hard to understand what `testAllowedVersion("0.12.0", "0.12.0+1", nextVersionType: NextVersionType.PATCH)` means. It's a little more confusing when looking at the parameter names in this case master is "0.12.0" and head is "0.12.0+1" is this a downgrade we're testing for?

If we're just testing `getAllowedVersions` what do you think about just testing it's public API, e.g:
```dart
Map<Version, NextVersionType> allowedVersions = getAllowedVersions("0.12.0");
expect(allowedVersions, {
"0.12.0+1": NextVersionType.PATCH,
"0.12.1": NextVersionType.MINOR,
"0.13.0": NextVersionType.BREAKING_MAJOR,
"1.0.0": NextVersionType.RELEASE
});
```

**Comment by iskakaushik** [comment]
Filed https://github.com/flutter/flutter/issues/35420 to track these pending comments. Will get to them soon.

**Pull request merged**